pkg/kubelet/volumes.go

```diff
@@ -137,33 +155,41 @@ func (kl *Kubelet) mountExternalVolumes(pod *api.Pod) (kubecontainer.VolumeMap,
 		// some volumes require attachment before mounter's setup.
 		// The plugin can be nil, but non-nil errors are legitimate errors.
 		// For non-nil plugins, Attachment to a node is required before Mounter's setup.
-		attacher, err := kl.newVolumeAttacherFromPlugins(volSpec, pod, volume.VolumeOptions{RootContext: rootContext})
+		attacher, err := kl.newVolumeAttacherFromPlugins(volSpec, pod)
 		if err != nil {
 			glog.Errorf("Could not create volume attacher for pod %s: %v", pod.UID, err)
 			return nil, err
 		}
 		if attacher != nil {
-			err = attacher.Attach(volSpec, kl.hostname)
-			if err != nil {
-				return nil, err
-			}
-
-			devicePath, err := attacher.WaitForAttach(volSpec, maxWaitForVolumeOps)
-			if err != nil {
+			// If the device path is already mounted, avoid an expensive call to the
+			// cloud provider.
+			deviceMountPath := attacher.GetDeviceMountPath(volSpec)
+			notMountPoint, err := kl.mounter.IsLikelyNotMountPoint(deviceMountPath)
+			if err != nil && !os.IsNotExist(err) {
 				return nil, err
 			}
-
-			deviceMountPath := attacher.GetDeviceMountPath(&volumeHost{kl}, volSpec)
-			if err = attacher.MountDevice(volSpec, devicePath, deviceMountPath, kl.mounter); err != nil {
-				return nil, err
+			if notMountPoint {
```

Comment:
Ok, that makes sense for GCE. What about volumes that don't do global mount, like NFS? Also, why is `mounter.SetUp(fsGroup)` called out side of this. Is it because we expect plugins that do Attach to all do a global mount?